[docs] Factory migration guide (#31095)
dagster-io/dagster: docs/docs/guides/build/components/asset-factories-to-components.md

<CodeExample
  path="docs_snippets/docs_snippets/guides/components/asset_factory/asset_factory.py"
  language="python"
  startAfter="start_build_etl_job"
  endBefore="end_build_etl_job"
/>

Within our component `asset_factory.py`, create a class that inherits from `dg.Model` with the attributes of the asset factory:

<CodeExample
  path="docs_snippets/docs_snippets/guides/components/asset_factory/asset_factory_component.py"
  language="python"
  title="src/<project_name>/components/asset_factory.py"
  startAfter="start_etl_job_model"
  endBefore="end_etl_job_model"
/>

Next we can use that Model within the `AssetFactory` class. At the top of the class create attributes for `access_key_id` and  `secret_access_key`. These will be shared across the assets and only need to be set once. The `etl_job` attribute will be a list since it can be any number of assets:

<CodeExample
  path="docs_snippets/docs_snippets/guides/components/asset_factory/asset_factory_component.py"
  language="python"
  title="src/<project_name>/components/asset_factory.py"
  startAfter="start_asset_factory_component"
  endBefore="end_asset_factory_component"
/>

> Most of `AssetFactory` code will look similar to the asset factory. Though the <PyObject section="definitions" module="dagster" object="Definitions" />  object returned contains all of the assets that will be generated as well as the resource.

Reviewer: ```suggestion
Most of the new `AssetFactory` code will look similar to the code in the old asset factory, although the <PyObject section="definitions" module="dagster" object="Definitions" />  object returned contains all of the assets that will be generated, as well as the resource.
```

> ### 3. Configure the component

Reviewer: ```suggestion
## 3. Use the component
```

> With the component registered, we can now use it in the project. The first step is to initialize the component:

Reviewer: ```suggestion
With the component created and registered, we can now use it in the project. The first step is to initialize the component:
```

<CliInvocationExample contents="dg scaffold defs 'my_project.components.asset_factory.AssetFactory' asset_factory" />

> Next set the attributes of the component:

Reviewer: ```suggestion
Next, set the attributes of the component:
```

<CodeExample
  path="docs_snippets/docs_snippets/guides/components/asset_factory/defs.yaml"
  language="yaml"
  title="src/<project_name>/defs/asset_factory/defs.yaml"
/>

> ### 4. Viewing component assets

Reviewer: ```suggestion
## 4. Viewing component assets
```

> The assets generated by the initialized component behave the same as those created by the factory. You can view them in the command line:

Reviewer: ```suggestion
The assets generated by the initialized component behave the same as those created by the factory. You can view them on the command line:
```

<CliInvocationExample contents="dg list defs" />

> Or interact with them in the Dagster UI by running `dg dev`.

Reviewer: ```suggestion
or interact with them in the Dagster UI by running `dg dev`:
```
